### ISSUE TITLE
TwitchLib with Owin.Security.Providers.Twitch

### ISSUE THREAD

**Issue by lexyoussef** [swiftyspiffy/TwitchLib#138]
Hi,

I wouldn't say that this is an issue, however I haven't found any results for best practices in configuring the Twitch client (specifically, the connection credentials), when using the Owin.Security.Providers.Twitch package, and I'm looking for some guidance in getting this started.

What's the secure way of grabbing the OAuth token to use for the connection credentials, and also being able to ensure that the token generated is using the right scope (let's just say for example that I want the scope to allow for all functionality).

tldr; Is there sample C# of using TwitchLib w/ Owin.Security.Providers.Twitch?

Thanks very much for your time and help!

P.S. Not new to C#, somewhat new to MVC

**Comment by swiftyspiffy** [comment]
Hey.

I am actually not very familiar with Own.Security.Providers.Twitch, but it looks like an OAuth implementation for ASP.NET?

The flow as I understand it is:
 - Contact Twitch with client application details.
 - Twitch returns a details that allow you to construct a link that you give to the user
 - User goes to the constructed URL and authorizes their account with Twitch
 - They get redirected to the redirect URL that you passed to Twitch
 - Extract token as GET param
 - Contact Twitch with this token and Twitch returns an access token
 - The access token is then used wherever you want to use their account.

The scopes list is a list that you build and put in the URL that you send the visitor to to get authorized. It's something like `user_read+channel_read+chat`.

I built a website that does the flow if you're interested in how it works. It's php, but the general idea is there. 

Site: https://twitchtokengenerator.com/
Source: https://github.com/swiftyspiffy/twitch-token-generator

In the mean time, I'll take a look at Owin.Security.Providers.Twitch.

**Comment by lexyoussef** [comment]
Thanks for your reply!

Yes that's correct, it's an OAuth impl. for ASP.NET.

I think where I'm a little lost is that I'm not sure how to control it such that I can pass in a custom scope, and then extract the token itself / where to intercept the authorization that Owin.Security.Providers.Twitch handles for you.

I'm downloading the twitchtokengenerator project now to get a better sense of the flow.

As for what I've done so far, it's pretty basic. Essentially a new MVC 4.5.2 ASP.NET project, installed Owin.Security.Providers.Twitch and TwitchLib. I've got the OAuth part working, its just finding that token once the authentication has taken place. 

Thanks again, I'll keep you posted should I make any progress!

**Comment by swiftyspiffy** [comment]
Just from a quick glance, the scopes are set in the TwitchAuthenticationOptions class as seen here:
https://github.com/TerribleDev/OwinOAuthProviders/blob/b2094b9c4fa5b65950ef4dd708b626f1f9908fbd/src/Owin.Security.Providers.Twitch/TwitchAuthenticationOptions.cs#L110

and TwitchAuthenticatedContext offers the access token:
https://github.com/TerribleDev/OwinOAuthProviders/blob/b2094b9c4fa5b65950ef4dd708b626f1f9908fbd/src/Owin.Security.Providers.Twitch/Provider/TwitchAuthenticatedContext.cs#L47

Not sure if this is helpful at all or not.

**Comment by lexyoussef** [comment]
Haha, yes, this helps tremendously! I guess I didn't know what the right questions were to ask in figuring this out.

Thank you very much for your help, I should be fine from here :)

**Comment by swiftyspiffy** [comment]
Awesome! :)
